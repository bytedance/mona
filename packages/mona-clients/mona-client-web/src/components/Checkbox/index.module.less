.checkbox {
  -webkit-tap-highlight-color: transparent;
  display: inline-block;

  & .wrapper {
    display: -webkit-inline-flex;
    display: inline-flex;
    -webkit-align-items: center;
    align-items: center;
    vertical-align: middle;
  }

  & .input {
    margin-right: 5px;
    -webkit-appearance: none;
    appearance: none;
    outline: 0;
    background-color: #FFF;
    border: 1px solid #E8E8E8;
    border-radius: 100%;
    width: 22px;
    height: 22px;
    position: relative;
  }

  & .input.checked:before {
    font: normal normal normal 14px/1 "ttui";
    content: "\e601";
    font-size: 28px;
    color: #FFFFFF;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -48%) scale(0.73);
    -webkit-transform: translate(-50%, -48%) scale(0.73);
  }
}