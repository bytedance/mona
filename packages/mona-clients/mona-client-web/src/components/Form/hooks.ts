import { useContext, useEffect, useRef } from "react";
import { FormContext } from './index';

export const useFormContext = (name: string, value: any, reset: () => void) => {
  const context = useContext(FormContext);
  const orderRef = useRef<number | null>(null);

  useEffect(() => {
    if (context) {
      const index = orderRef.current;
      orderRef.current = context.register(
        () => ({ name, value: typeof value === 'function' ? value() : value }),
        () => reset(),
        index
      )

      return () => {
        context.unregister(index);
      }
    }
    return () => {}
  }, [context, name, value, reset])
}