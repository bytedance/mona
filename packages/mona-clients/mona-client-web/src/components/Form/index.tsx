import React from 'react';
import { FormProps } from '@bytedance/mona';

const Form: React.FC<FormProps> = ({ children }) => {
  return <div>{children}</div>
}

export default Form;
