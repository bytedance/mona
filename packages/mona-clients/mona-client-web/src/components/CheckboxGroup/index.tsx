import React, { useRef } from 'react';
import { CheckboxGroupProps, TouchEvent } from '@bytedance/mona';
import { useHandlers } from '../hooks';

export const EMPTY_ITEM = Symbol('checkboxEmpty');

interface CheckboxGroupContextProps {
  initValue: (value?: string, checked?: boolean) => number;
  changeValue: (index: number, value?: string) => void
  toggleChecked: (index: number, checked?: boolean, e?: TouchEvent) => void
}

interface ValueItem {
  value?: string;
  checked?: boolean;
}

export const CheckboxGroupContext = React.createContext<CheckboxGroupContextProps | null>(null)

const CheckboxGroup: React.FC<CheckboxGroupProps> = ({ children, onChange, ...restProps }) => {
  const { handleClassName, ...handlerProps } = useHandlers(restProps);
  // const [values, setValues] = useState<ValueItem[]>([]);
  const valuesRef = useRef<ValueItem[]>([])
  const indexRef = useRef(0);
  
  const context = {
    initValue: (value?: string, checked?: boolean) => {
      const index = indexRef.current++;
      const newValues = [...valuesRef.current];
      // record order
      newValues[index] = { value, checked };
      valuesRef.current = newValues;
      return index;
    },
    changeValue: (index: number, value?: string) => {
      const newValues = [...valuesRef.current];
      newValues[index] = { ...newValues[index], value };
      valuesRef.current = newValues;
    },
    toggleChecked: (index: number, checked?: boolean, event?: TouchEvent) => {
      const newValues = [...valuesRef.current];
      newValues[index] = { ...newValues[index], checked }
      valuesRef.current = newValues;
      
      if (typeof onChange === 'function') {
        const result = newValues.filter(v => v.checked).map(v => v.value);
        const e = { ...event!, detail: { value: result as string[] } };
        onChange(e)
      }
    }
  }

  return (
    <div className={handleClassName()} {...handlerProps}>
      <CheckboxGroupContext.Provider value={context}>
        {children}
      </CheckboxGroupContext.Provider>
    </div>
  )
}

export default CheckboxGroup;
