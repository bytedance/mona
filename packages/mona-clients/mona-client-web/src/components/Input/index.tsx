import React from 'react';
import { InputProps } from '@bytedance/mona';
import styles from './index.module.less';
import { useHandlers } from '../hooks';

const Input: React.FC<InputProps> = ({ children, ...restProps }) => {
  const { handleClassName, ...handlerProps} = useHandlers(restProps)
  
  return <input className={handleClassName(styles.input)} {...handlerProps} />
}

export default Input;
