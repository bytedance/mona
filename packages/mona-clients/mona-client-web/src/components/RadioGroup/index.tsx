import React, { useCallback, useRef } from 'react';
import { RadioGroupProps, TouchEvent } from '@bytedance/mona';
import { useHandlers } from '../hooks';
import { useFormContext } from '../Form/hooks';

interface RadioGroupContextProps {
  initValue: (value?: string, checked?: boolean, clear?: () => void) => number;
  changeValue: (index: number, value?: string) => void
  toggleChecked: (index: number, checked?: boolean, e?: TouchEvent) => void
}

interface ValueItem {
  value?: string;
  checked?: boolean;
}

export const RadioGroupContext = React.createContext<RadioGroupContextProps | null>(null)

const RadioGroup: React.FC<RadioGroupProps> = ({ children, name = '', onChange, ...restProps }) => {
  const { handleClassName, ...handlerProps } = useHandlers(restProps);
  const valuesRef = useRef<ValueItem[]>([])
  const indexRef = useRef(0);
  const clearsRef = useRef<((() => void) | undefined)[]>([]);

  const reset = useCallback(() => {
    const newValues = valuesRef.current;
    const cbs = clearsRef.current;
    for (let i = 0; i < newValues.length; i++) {
      newValues[i] = { ...newValues[i], checked: false }
      const cb = cbs[i];
      if (typeof cb === 'function') {
        cb();
      }
    }
  }, [clearsRef, valuesRef]);
  const getValues = () => {
    return valuesRef.current.filter(v => v.checked).map(v => v.value)[0] || '';
  }
  useFormContext(name, getValues, reset);
  
  const context = {
    initValue: (value?: string, checked?: boolean, clear?: () => void) => {
      const index = indexRef.current++;
      const newValues = [...valuesRef.current];
      // record order
      newValues[index] = { value, checked };
      valuesRef.current = newValues;
      clearsRef.current[index] = clear;
      return index;
    },
    changeValue: (index: number, value?: string) => {
      const newValues = [...valuesRef.current];
      newValues[index] = { ...newValues[index], value };
      valuesRef.current = newValues;
    },
    toggleChecked: (index: number, checked?: boolean, event?: TouchEvent) => {
      const newValues = [...valuesRef.current];
      const cbs = clearsRef.current;

      for (let i = 0; i < newValues.length; i++) {
        if (index !== i) {
          newValues[i] = { ...newValues[i], checked: false }
          const cb = cbs[i];
          if (checked && typeof cb === 'function') {
            cb();
          }
        } else {
          newValues[i] = { ...newValues[i], checked }
        }
      }

      valuesRef.current = newValues;
      
      if (typeof onChange === 'function') {
        const value = newValues[index];
        const e = { ...event!, detail: { value: value.checked ? value.value : undefined } };
        onChange(e)
      }
    }
  }

  return (
    <div className={handleClassName()} {...handlerProps}>
      <RadioGroupContext.Provider value={context}>
        {children}
      </RadioGroupContext.Provider>
    </div>
  )
}

export default RadioGroup;
