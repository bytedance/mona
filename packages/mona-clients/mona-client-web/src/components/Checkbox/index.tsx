import React, { useState, useEffect, useContext, useRef } from 'react';
import { CheckboxProps } from '@bytedance/mona';
import cs from 'classnames';
import { CheckboxGroupContext } from '../CheckboxGroup';
import styles from './index.module.less';
import { useHandlers } from '../hooks';
import { formatMouseEvent } from '../utils';

const Checkbox: React.FC<CheckboxProps> = (props) => {
  const {
    value,
    disabled,
    checked,
    color = '#F85959',
    children,
    id,
    ...restProps
  } = props;

  const group = useContext(CheckboxGroupContext);
  const orderRef = useRef(-1);

  const { handleClassName, onClick, ...handlerProps } = useHandlers(restProps);

  const [isChecked, setIsChecked] = useState(checked);

  const handleClick = (e: React.MouseEvent) => {
    const newIsChecked = !isChecked;
    setIsChecked(newIsChecked);
    if (orderRef.current >= 0) {
      group?.toggleChecked(orderRef.current, newIsChecked, formatMouseEvent({ event: e, type: 'change' }));
    }
    onClick(e);
  }
  
  // don't change order of useEffects, the order is important
  useEffect(() => {
    setIsChecked(checked);
    if (orderRef.current >= 0) {
      group?.toggleChecked(orderRef.current, checked);
    }
  }, [checked])

  useEffect(() => {
    orderRef.current = group ? group?.initValue(value, checked) : -1;
  }, [])

  useEffect(() => {
    if (orderRef.current >= 0) {
      group?.changeValue(orderRef.current, value);
    }
  }, [value])

  return (
    <div
      className={handleClassName(styles.checkbox)}
      {...handlerProps}
    >
      <div className={styles.wrapper}>
        <div className={cs(styles.input, {[styles.checked]: isChecked})} style={{ borderColor: color, backgroundColor: isChecked ? color: '' }}>
          <input value={value} id={id} onClick={handleClick} type="checkbox" className={styles.inner}></input>
        </div>
        {children}
      </div>
    </div>
  )
}

export default Checkbox;
