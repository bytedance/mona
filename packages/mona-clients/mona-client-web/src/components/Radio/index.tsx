import React, { useState, useEffect, useContext, useRef } from 'react';
import { RadioProps } from '@bytedance/mona';
import { RadioGroupContext } from '../RadioGroup';
import styles from './index.module.less';
import { useHandlers } from '../hooks';
import { formatMouseEvent } from '../utils';

const Radio: React.FC<RadioProps> = props => {
  const { value, disabled, checked, color = '#F85959', children, id, ...restProps } = props;

  const group = useContext(RadioGroupContext);
  const orderRef = useRef(-1);

  const { handleClassName, onClick, ...handlerProps } = useHandlers(restProps);

  const [isChecked, setIsChecked] = useState(checked);

  const handleClick = (e: React.MouseEvent) => {
    const newIsChecked = !isChecked;
    setIsChecked(newIsChecked);
    if (orderRef.current >= 0) {
      group?.toggleChecked(orderRef.current, newIsChecked, formatMouseEvent({ event: e, type: 'change' }));
    }
    onClick(e);
  };

  // don't change order of useEffects, the order is important
  useEffect(() => {
    setIsChecked(checked);
    if (orderRef.current >= 0) {
      group?.toggleChecked(orderRef.current, checked);
    }
  }, [checked]);

  useEffect(() => {
    orderRef.current = group ? group?.initValue(value, checked, () => setIsChecked(false)) : -1;
  }, []);

  useEffect(() => {
    if (orderRef.current >= 0) {
      group?.changeValue(orderRef.current, value);
    }
  }, [value]);

  return (
    <div className={handleClassName(styles.radio)} {...handlerProps}>
      <div className={styles.wrapper}>
        <div
          className={`${styles.input} ${isChecked ? styles.checked : ''}`}
          style={{ borderColor: color, backgroundColor: isChecked ? color : '' }}
        >
          <input id={id} value={value} onClick={handleClick} type="radio" className={styles.inner}></input>
        </div>
        {children}
      </div>
    </div>
  );
};

export default Radio;
