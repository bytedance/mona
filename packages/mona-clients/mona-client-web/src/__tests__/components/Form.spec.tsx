import React from 'react';
import { mount } from 'enzyme';
import Form from '../../components/Form';
import mountTest from '../../../../../tests/shared/mountTest';
import handlerTest from '../../../../../tests/shared/handlerTest';
import { FormProps } from '@bytedance/mona';
import Checkbox from '../../components/Checkbox';
import Input from '../../components/Input';
import CheckboxGroup from '../../components/CheckboxGroup';
import RadioGroup from '../../components/RadioGroup';
import Radio from '../../components/Radio';

describe('web component: Form', () => {
  mountTest(Form)
  handlerTest({ Component: Form, firstTag: 'form' });

  it('should render correctly', () => {
    expect(mount(
      <Form>
        <Input />
        <Checkbox />
      </Form>
    ).render()).toMatchSnapshot();
  })
  
  it('should have right properties', () => {
    const props: FormProps = {
      onReset: () => {},
      onSubmit: () => {},
      className: 'class-name',
      style: { color: 'red' },
    }
    const wrapper = mount(<Checkbox {...props}>hello</Checkbox>);
    expect(wrapper.render()).toMatchSnapshot();
  })

  it('should trigger onSubmit and onReset correctly', () => {
    const onSubmit = jest.fn();
    const onReset = jest.fn();
    const wrapper = mount(
      <Form onSubmit={onSubmit} onReset={onReset}>
        <Input name="input" value="helloworld"></Input>
        <CheckboxGroup name="checkbox">
          <Checkbox value="apple">apple</Checkbox>
          <Checkbox value="banana">banana</Checkbox>
        </CheckboxGroup>
        <RadioGroup name="radio">
          <Radio value="black">red</Radio>
          <Radio value="red">red</Radio>
        </RadioGroup>
      </Form>
    )

    wrapper.find('input[value="apple"]').first().simulate('click');
    wrapper.find('input[value="black"]').first().simulate('click');
    wrapper.find('form').simulate('submit');
    expect(onSubmit).toBeCalledTimes(1);
    expect(onSubmit.mock.calls[0][0].detail.value).toEqual({ input: 'helloworld', checkbox: ['apple'], radio: 'black' })
    wrapper.find('form').simulate('reset');
    expect(onReset).toBeCalledTimes(1);
    wrapper.find('form').simulate('submit');
    expect(onSubmit.mock.calls[1][0].detail.value).toEqual({ input: '', checkbox: [], radio: '' })
    // TODO: add more component
  })
})
