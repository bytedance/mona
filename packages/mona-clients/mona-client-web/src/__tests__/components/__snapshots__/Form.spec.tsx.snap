// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`web component: Form should have right properties 1`] = `
LoadedCheerio {
  "0": <div
    class="checkbox class-name  "
    style="color: red;"
  >
    <div
      class="wrapper"
    >
      <div
        class="input "
        style="border-color: #f85959;"
      >
        <input
          class="inner"
          type="checkbox"
          value=""
        />
      </div>
      hello
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`web component: Form should render correctly 1`] = `
LoadedCheerio {
  "0": <form
    class="   "
  >
    <input
      class="input   "
      enterkeyhint="done"
      maxlength="140"
      name=""
      type="text"
      value=""
    />
    <div
      class="checkbox   "
    >
      <div
        class="wrapper"
      >
        <div
          class="input "
          style="border-color: #f85959;"
        >
          <input
            class="inner"
            type="checkbox"
            value=""
          />
        </div>
      </div>
    </div>
  </form>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
